src/main/java/org/opensearch/ad/transport/PMMLResultRequest.java

```diff
@@ -0,0 +1,148 @@
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package org.opensearch.ad.transport;
+
+import static org.opensearch.action.ValidateActions.addValidationError;
+
+import java.io.IOException;
+
+import org.opensearch.action.ActionRequest;
+import org.opensearch.action.ActionRequestValidationException;
+import org.opensearch.ad.constant.CommonErrorMessages;
+import org.opensearch.ad.constant.CommonName;
+import org.opensearch.common.Strings;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.common.xcontent.ToXContentObject;
+import org.opensearch.common.xcontent.XContentBuilder;
+
+public class PMMLResultRequest extends ActionRequest implements ToXContentObject {
+    private String adID;
+    private String mlModelID;
+    private String[] featureNames;
+    private double[] features;
+
+    // Messages used for validation error
+    public static final String INVALID_FEATURE_NAME_MSG = "feature name vector is empty (in pmml result request)";
+    public static final String INVALID_FEATURE_MSG = "feature vector is empty (in pmml result request)";
+    public static final String INVALID_LENGTH_MSG = "feature name vector and feature vector have different lengths";
+
+    public PMMLResultRequest(String adID, String mlModelID, String[] featureNames, double[] features) {
+        super();
+        this.adID = adID;
+        this.mlModelID = mlModelID;
+        this.featureNames = featureNames;
+        this.features = features;
+    }
+
+    public PMMLResultRequest(StreamInput in) throws IOException {
+        super(in);
+        adID = in.readString();
+        mlModelID = in.readString();
+        int size1 = in.readVInt();
+        featureNames = new String[size1];
+        for (int i = 0; i < size1; i++) {
+            featureNames[i] = in.readString();
+        }
+        int size2 = in.readVInt();
+        features = new double[size2];
+        for (int i = 0; i < size2; i++) {
+            features[i] = in.readDouble();
+        }
+    }
+
+    public String getAdID() {
+        return adID;
+    }
+
+    public String getMlModelID() {
+        return mlModelID;
+    }
+
+    public String[] getFeatureNames() {
+        return featureNames;
+    }
+
+    public double[] getFeatures() {
+        return features;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        out.writeString(adID);
+        out.writeString(mlModelID);
+        out.writeVInt(featureNames.length);
+        for (String name : featureNames) {
```

Comment:
Use `out.writeOptionalArray` ?

Reply:
I used this way to stick to how it's done in the RCF result request. I will dm you to discuss some details about this.

Reply:
From our discussion, `writeOptionalArray` can only work for `Writeable` class, but String is not. It's ok to keep as is.